lib/grizzly/zwave_firmware.ex

```diff
@@ -0,0 +1,167 @@
+          )
+
+      {:error, :firmware_dir_does_not_exist} ->
+        Logger.info(
+          "[Grizzly]: firmware directory does not exist, not attempting a firmware update"
+        )
+
+      {:error, :zw_programmer_does_not_exist} ->
+        Logger.info(
+          "[Grizzly]: zw_programmer binary does not exist, not attempting a firmware update"
+        )
+
+      {:error, :no_zwave_firmware_file} ->
+        Logger.info("[Grizzly]: could not detect firmware file, not attempting a firmware update")
+    end
+
+    :ok
+  end
+
+  defp check_firmware_dir(%{zwave_firmware_dir: nil}) do
+    {:error, :no_firmware_dir}
+  end
+
+  defp check_firmware_dir(options) do
+    Logger.info("[Grizzly] Checking for firmware dir #{inspect(options.zwave_firmware_dir)}")
+
+    if File.exists?(firmware_dir(options)) do
+      :ok
+    else
+      {:error, :firmware_dir_does_not_exist}
+    end
+  end
+
+  defp firmware_dir(options), do: Application.app_dir(:grizzly, options.zwave_firmware_dir)
+
+  defp check_zwave_programmer_exists(options) do
+    if File.exists?(options.zw_programmer_path) do
+      :ok
+    else
+      {:error, :zw_programmer_does_not_exist}
+    end
+  end
+
+  defp get_firmware_file(options) do
+    dir = firmware_dir(options)
+
+    case File.ls!(dir) do
+      [file] ->
+        {:ok, Path.join(dir, file)}
+
+      _ ->
+        {:error, :no_zwave_firmware_file}
+    end
+  end
+
+  defp maybe_upload_zwave_firmware_update(gbl_file, options) do
+    with {:ok, gbl_version} <- gbl_version(gbl_file),
+         {:ok, current_version} <- current_version(options),
+         :ok <- validate_upgrade(current_version, gbl_version) do
+      _ = upload_zwave_firmware_update(gbl_file, options)
+    else
+      {:error, reason} ->
+        Logger.info("[Grizzly] Passed on uploading new zwave firmware: #{reason}")
+    end
+
+    :ok
+  end
+
+  # .../ZW_SerialAPI_Controller_7.16.3_297_ZGM130S_REGION_US_LR.gbl
+  defp gbl_version(gbl_file) do
+    basename = Path.basename(gbl_file)
+    regex = ~r/\w+_(?<major>\d+)\.(?<minor>\d+)\.(?<patch>\d+).*\.gbl/
+
+    Regex.named_captures(regex, basename)
+    |> parse_version_capture()
+  end
+
+  defp current_version(options) do
+    with {answer, 0} <- System.cmd(options.zw_programmer_path, ["-s", options.serial_port, "-t"]),
+         {:ok, current_version} <- extract_current_version(answer) do
+      {:ok, current_version}
+    else
+      other ->
+        Logger.warn(
+          "[Grizzly] Failed to get the hub's current firmware version: #{inspect(other)}"
+        )
+
+        {:error, :invalid}
+    end
+  end
+
+  # "Using serial device /dev/ttyS4\nConnected to Serial device: OK\nSerial version: 9, Chip type: 7, Chip version: 0, SDK: 7.15.02, ..."
+  defp extract_current_version(answer) do
+    Logger.info("[Grizzly] Extracting current version from #{inspect(answer)}")
+    regex = ~r/.*SDK:\s*(?<major>\d+)\.(?<minor>\d+)\.(?<patch>\d+)/m
+
+    Regex.named_captures(regex, answer)
```

Comment:
Any reason not to use Elixir's `Version` module to parse and compare version numbers?

Reply:
Now using Version to compare versions. But it can't be used to parse since it fails on "7.15.02" which zw_programmer returns.